streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -901,6 +901,36 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
   )(f: (B, C) => D)(implicit ev: A00 =:= A1): ZSink[R1, E1, A00, A1, D] =
     zip(that).map(f.tupled)
 
+  final def andThen[R1 <: R, E1 >: E, A00 >: A0, C, A000](that: ZSink[R1, E1, A00, B, C]): ZSink[R1, E1, A0, A, C] =
+    new ZSink[R1, E1, A0, A, C] {
+      override type State = that.State
+
+      /**
+       * Runs the sink from an initial state and produces a final value
+       * of type `B`
+       */
+      override def extract(state: State): ZIO[R1, E1, C] = that.extract(state)
+
+      /**
+       * The initial state of the sink.
+       */
+      override def initial: ZIO[R1, E1, Step[State, Nothing]] =
+        that.initial
+
+      /**
+       * Steps through one iteration of the sink
+       */
+      override def step(state: State, a: A): ZIO[R1, E1, Step[State, A0]] =
+        for {
+          s0 <- self.initial
+          s1 <- self.step(Step.state(s0), a)
+          b  <- self.extract(Step.state(s1))
```

Comment:
You'll need to check if `this` is actually done here, before extracting. Otherwise this may fail for sinks that do not allow for partial extracts.

Reply:
> Hey @dkarlinsky, thanks for picking this up! I thought this through for a bit, and I think that the only viable strategy is to not allow for leftovers in `that`. This is not to say that we should require `A00 = Nothing`, because that would be too limiting. But rather, we should always signal `Step.more` when `that` has leftovers, and we should buffer those leftovers and re-step them when more input can be extracted from `this`.
> 
> Does that make sense?

I think I understand the general idea, but I maybe be missing something:
If I buffer leftovers for `that`, which is of type `Chunk[A00]`, how do I try to feed them back into next `that.step()`, which expects type `B`?